internal/wrr/random.go

```diff
@@ -51,27 +53,40 @@ var grpcrandInt63n = grpcrand.Int63n
 func (rw *randomWRR) Next() (item interface{}) {
 	rw.mu.RLock()
 	defer rw.mu.RUnlock()
-	if rw.sumOfWeights == 0 {
+	var sumOfWeights int64
+	if len(rw.items) == 0 {
 		return nil
 	}
-	// Random number in [0, sum).
-	randomWeight := grpcrandInt63n(rw.sumOfWeights)
-	for _, item := range rw.items {
-		randomWeight = randomWeight - item.Weight
-		if randomWeight < 0 {
-			return item.Item
-		}
+	if rw.equalWeights {
+		return rw.items[grpcrandInt63n(int64(len(rw.items)))].item
 	}
-
-	return rw.items[len(rw.items)-1].Item
+	sumOfWeights = rw.items[len(rw.items)-1].accumulatedWeight
+	// Random number in [0, sumOfWeights).
+	randomWeight := grpcrandInt63n(sumOfWeights)
+	// Item's accumulated weights are in ascending order, because item's weight >= 0.
+	// Binary search rw.items to find first item whose accumulatedWeight > randomWeight
+	// The return i is guaranteed to be in range [0, len(rw.items)) because randomWeight < last item's accumulatedWeight
+	i := sort.Search(len(rw.items), func(i int) bool { return rw.items[i].accumulatedWeight > randomWeight })
+	return rw.items[i].item
 }
 
 func (rw *randomWRR) Add(item interface{}, weight int64) {
 	rw.mu.Lock()
 	defer rw.mu.Unlock()
-	rItem := &weightedItem{Item: item, Weight: weight}
+	accumulatedWeight := weight
+	equalWeights := true
+	if len(rw.items) > 0 {
+		lastItem := rw.items[len(rw.items)-1]
+		accumulatedWeight = lastItem.accumulatedWeight + weight
+		lastItemWeight := lastItem.accumulatedWeight
+		if len(rw.items) > 1 {
```

Comment:
Oh, I now see what you meant in the old comments. `weight` is used once here.
I would be OK to keep the weight field to simplify the code here.
Either way is fine.

Reply:
Added  `weight` back to `weightedItem` to simplify code here

Reply:
By the way, why `weight` type is `int64` not `uint32`,  `int64` will allow user add negative weight item to `WRR` which will result undefined behaviour in `Next`.